tests.texlua/memoize-extract.lua

```diff
@@ -0,0 +1,1279 @@
+		local pdf_version = ("%d.%d"):format(v_major, v_minor)
+
+		pdfe.close(pdf)
+
+		table.sort(succ)
+		table.sort(failed, function(a,b) return a.page.page < b.page.page end)
+		return succ, failed, pdf_version
+	end
+end
+
+-- restrict the complete rest of the script by undefining security relevant libraries
+-- this defines an allow-list what functions of these libraries still should be accessible
+local env = {
+	-- lua libraries
+	arg      = arg,
+	ipairs   = ipairs,
+	math     = math,
+	os       = { type = os.type, },
+	pairs    = pairs,
+	print    = print,
+	table    = table,
+	tonumber = tonumber,
+	tostring = tostring,
+	select   = select,
+
+	-- luatex specific libraries
+	lfs      = {isfile=lfs.isfile},
+	kpse     = kpse,
+
+	-- memoize-extract specific global
+	STAGE    = STAGE,
+}
+
+local exit
+if STAGE == "testing" then
+	-- in testing environment avoid exiting the whole test
+	-- -> instead raise an error which can be catched
+
+	-- store the error function independent of the environment
+	local error = error
+	exit = {
+		error = function() error("exited with error") end,
+		warn  = function() error("exited with warn") end,
+		succ  = function() error("exited with succ") end,
+	}
+else
+	local os_exit = os.exit
+	exit = {
+		error = function() os_exit(11) end,
+		warn  = function() os_exit(10) end,
+		succ  = function() os_exit(0) end,
+	}
+end
+
+do
+	-- I don't like using the debug library, but getting a traceback here is a
+	-- must to find where the error originates from
+	local debug_traceback = debug.traceback
+	-- use the lua error function -> exits immediately
+	local error = error
+
+	-- Prevent trying to change the environment.
+	local function bad_index(...)
+		local msg = "Attempt to access an undefined index:"
+		for i = 2, select("#", ...) do
+			msg = msg ..tostring(select(i, ...)).." "
+		end
+		msg = msg.."\n\n"..debug_traceback(nil, 2)
+		error(msg)
+	end
+	setmetatable(env, {
+		__index     = bad_index,
+		__metatable = false,
+		__newindex  = bad_index,
+	})
+end
+
+_ENV = env
+----------------------------------
+-- restricted area startes here --
+----------------------------------
+
+---------------
+-- pathutils --
+---------------
+-- -> probably moved to a different library eventually
+local pathlib = {}
+do
+	-- other projects like penlight or l3build do stuff with different pathseps, according to
+	-- https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#canonicalize-separators
+	-- and
+	-- https://retrocomputing.stackexchange.com/questions/28344/since-when-does-windows-support-forward-slash-as-path-separator
+	-- windows since quite a while also works with / as pathsep.
+	-- Thus, this pathlib will normalize paths for having / as pathsep before working on paths
+	if os.type == "windows" then
+		---Normalize path such that windows also uses /
+		---@param path string
+		---@return string
+		function pathlib.path_normalize(path)
+			return path:gsub("\\", "/")
+		end
```

Comment:
So if `memoize-extract.lua` will only ever be ran via shell-escape/`\write18`, you probably don't even need this, since kpathsea and most TeX packages exclusively use `/` as the path separator. There's nothing wrong with this from a security perspective though, so there's no problem keeping this if you want to support Windows users who intentionally use `\` path separators.

Reply:
Most often it will be run from TeX directly, but for example I like to run the script manually outside of the context of TeX (just like e.g. `biber` or so). So for compatibility it's probably better to keep it (except we run in too many issues with windows paths which would be a reason for me to drop this "feature")